【功能模块开发】外部交互模块

revue-starlight opened the pull request:
不太会写，会继续写。dbq

Commit:
frame

Commit:
reqcat

Commit:
210209

revue-starlight committed:
Merge pull request #1 from ThinkSpiritLab/master

更新

Commit:
new commit

Commit:
newcommit

Commit:
externalmodule

Commit:
externalmodule

Commit:
externalmodule

Commit:
externalmodule

Commit:
externalmodule

Commit:
0

Commit:
externalmodule

Commit:
externalmodule

Commit:
externalmodule

Commit:
externalmodule

A0nameless0man commented on .eslintrc.js:
```diff
@@ -20,7 +20,7 @@ module.exports = {
     rules: {
         "indent": ["error", 4, { "SwitchCase": 1 }],
-        "linebreak-style": ["error", "unix"],
+        "linebreak-style": ["off", "windows"],
```
不要瞎搞，vscode有改变换行模式的选项

revue-starlight replied:
嗯，修改了

A0nameless0man commented on src/external-module/external-module.controller.ts:
```diff
@@ -0,0 +1,17 @@
+    ) {}
+
+    @Post("v1/judgers/token")
+    async JudgeLogin(@Req() request: any): Promise<any> {
```
外部模块不负责judger啊……

revue-starlight replied:
这里的原来的想法是评测端用http post登录,然后返回token，然后建立连接。是不是有啥问题....

A0nameless0man replied:
确实是这个流程，但是这部分工作是由内部模块负责处理的。

A0nameless0man commented on src/external-module/external-module.service.ts:
```diff
@@ -0,0 +1,104 @@
+
+    // 发送评测任务并等待回调
+    async createjudge(req: CreateJudgeRequest): Promise<void> {
+        if ((await this.redisService.client.get("Taskids")) == null)
```
redis对于整数的操作提供不少原子方法，请看文档。

revue-starlight replied:
谢谢qaqaqaqaqaqaq

A0nameless0man replied:
又及，这是不是某种全局自增主键？

revue-starlight replied:
嗯，已经将自增改为client.incr()。不过redis的值好像只能存储字符串

Handgrip replied:
JudgeLogin Exit getLog reportStatus createjudge这几个函数都和内部通信重复了。外部通信只负责和oj的web端通信，只用于和web端收发任务，回报任务状态；不负责管理/操纵评测机，不负责和评测机通信。
另外各模块的管理员接口是分开提供的吗
@Nugine 
另外收到任务之后，taskid要自动生成吗，还是用web端提供的taskid。 @A0nameless0man @Nugine

revue-starlight replied:
web端好像没有提供taskid，应该是外部模块这里提供

Nugine replied:
管理接口分开提供。

外部系统发的请求中携带了id，外部通信模块需要对这个输入的id做重新映射，生成一个内部使用的taskId，以免不同的外部系统之间发生冲突，给外部系统发回调时还是用原来输入的id。

https://github.com/ThinkSpiritLab/Heng-Protocol/blob/239ab32b44e5531094b37b5262ca5f573d93a11a/external-protocol/index.ts#L19

A0nameless0man replied:
按照我的理解，外部系统只要自己把id编码到回调url里即可。发来的请求不需要包含id

Nugine replied:
这样也可以，先按内部生成taskId写，忽略有关外部id的内容。

A0nameless0man commented on src/external-module/external-module.service.ts:
```diff
@@ -0,0 +1,104 @@
+        console.log("发送id为: ${taskid} 的任务");
+        await this.judgequeueService.push(String(taskid));
+        //塞入队列后，等待评测机评测完毕后将结果回调
+        //Q:评测机直接调用url还是在这里调用呢？
```
提供一个方法供内部模块调用

revue-starlight replied:
嗯嗯

A0nameless0man replied:
具体而言，评测机会通过wsrpc通知控制端。内部模块收到通知后，会调用外部模块提供的方法。外部模块进行实际的HTTP回调。

A0nameless0man commented on tsconfig.json:
```diff
@@ -13,7 +13,7 @@
         // "outFile": "./",                       /* Concatenate and emit output to single file. */
         "outDir": "./dist", /* Redirect output structure to the directory. */
-        // "rootDir": "./",                       /* Specify the root directory of input files. Use to control the output directory structure with --outDir. */
+        "rootDir": "./",                       /* Specify the root directory of input files. Use to control the output directory structure with --outDir. */
```
为啥改这个？

revue-starlight replied:
这个之前注释掉会谜之报错。。。但是现在好像又好了

revue-starlight replied:
> 为啥改这个？

找到报的错了，在import { 一些东西 } from “src/一些东西" 时，路径"src/一些东西"会报错

A0nameless0man commented on yarn.lock:
```diff
@@ -561,10 +561,17 @@
     webpack "5.9.0"
     webpack-node-externals "2.5.2"
 
+<<<<<<< HEAD
```
冲突了，兄弟……

revue-starlight replied:
嗯..

Nugine commented:
我们之前没有具体规定外部交互模块的行为，作者可以先写一下这个PR做了哪些改进，方便讨论。

没有完成的PR可以转换为 draft 模式，自动测试一样会运行，reviewer 也可以看到代码。完成后可以退出draft模式。

请不要改动编译配置和格式化配置，除非发生了什么冲突或者特殊问题。如果有，要在PR中说明一下。

Nugine commented:
目前[外部通信协议](https://github.com/ThinkSpiritLab/Heng-Protocol/blob/master/docs/external-protocol.md)定义了两个接口。PR 作者需要实现这两个接口。

内部通信模块要在某个 service 中提供读取评测端状态的方法。

外部通信模块增加一个 service，实现：
1. 读取控制端状态
2. 保存评测任务信息，把任务ID传递给评测队列
3. 提供更新任务状态的方法，由内部通信模块调用。该方法被调用时会向外部系统给出的回调url发请求，通知任务状态发生变化。
4. 提供完成任务的方法，由内部通信模块调用。该方法被调用时会向外部系统给出的回调url发请求，通知任务的最终结果。

增加一个 controller，实现：
1. 查询系统负载接口：用对应的 service 取得控制端和评测端的状态信息，返回状态信息
2. 创建评测任务接口：校验请求体格式，用对应的 service 把评测任务传递给评测队列，返回这个任务的ID

@revue-starlight @Flaryer 你们讨论一下怎么对接

Handgrip commented:
外部通信还需要两个接口：汇报任务状态和任务结果（由内部通信调用，调用后将状态或者结果发送给给定的url）。具体见下图。
![process](https://user-images.githubusercontent.com/30646834/108058245-b63c3c80-708e-11eb-8f3e-37dbb0e51f46.gif)


另外获取评测机状态和控制端状态是否属于管理员接口了。 @Nugine

Nugine commented:
> 外部通信还需要两个接口：汇报任务状态和任务结果（由内部通信调用，调用后将状态或者结果发送给给定的url）。具体见下图。
> ![process](https://user-images.githubusercontent.com/30646834/108058245-b63c3c80-708e-11eb-8f3e-37dbb0e51f46.gif)
> 
> 另外获取评测机状态和控制端状态是否属于管理员接口了。 @Nugine

查询状态属于管理接口，可以拆到另一个 controller

A0nameless0man commented:
管理接口好像一开始说由各个模块自行提供，不进行集中。
另外，有人记得Guard模块谁来写吗。

Handgrip commented:
我记得Guard还没人接锅

revue-starlight commented:
嗯...------------------&nbsp;原始邮件&nbsp;------------------
发件人:&nbsp;&quot;Flaryer&quot;<notifications@github.com&gt;
发送时间:&nbsp;2021年2月17日(星期三) 上午7:37
收件人:&nbsp;&quot;ThinkSpiritLab/heng-controller&quot;<heng-controller@noreply.github.com&gt;;
抄送:&nbsp;&quot;revue-starlight&quot;<204214884@qq.com&gt;;&quot;Mention&quot;<mention@noreply.github.com&gt;;
主题:&nbsp;Re: [ThinkSpiritLab/heng-controller] 外部交互模块（没有写完） (#17)

revue-starlight committed:
UPDATE

Commit:
upd

revue-starlight committed:
upd

revue-starlight committed:
upd

revue-starlight committed:
[外部控制模块]部分通信

revue-starlight committed:
[外部控制模块]部分通信

revue-starlight committed:
[外部控制模块]部分通信

revue-starlight committed:
Merge pull request #2 from revue-starlight/cyq

[外部通信模块] 基本http通信和回调

revue-starlight committed:
[外部通信模块]

实现了与内部通信模块的对接
删除了package-lock.json
修改了readme文档的地址

revue-starlight committed:
外部通信模块

更改一些文件的位置

revue-starlight committed:
[外部交互模块]

封装了键名
jduger模块改变取judgeINFO 方式为获取直接对象

revue-starlight committed:
[external-module]

封装了redis键
将judger模块中的getJUDGEINFO从外部模块直接返回对象

revue-starlight committed:
[外部控制模块] 改变judgeinfo传递格式

Handgrip commented on src/judger/judger.service.ts:
```diff
@@ -158,7 +148,10 @@ export class JudgerService {
             );
 
-        // TODO 通知外部系统
+        this.externalmoduleService.responsefinish(
```
await

Handgrip commented on src/judger/judger.service.ts:
```diff
@@ -137,7 +124,10 @@ export class JudgerService {
                 `回报无效任务状态 ${args.length - vaildResult.length} 个`
             );
-        // TODO 通知外部系统
+        this.externalmoduleService.responseupdate(
```
await

Handgrip commented on src/judger/judger.service.ts:
```diff
@@ -158,7 +148,10 @@ export class JudgerService {
-        // TODO 通知外部系统
+        this.externalmoduleService.responsefinish(
+            parseInt(args[0].id),
+            vaildResult
```
上面solveUpdateJudges()也是这个情况：
vaildResult的类型：
```ts
const vaildResult: {
    id: string;
    result: JudgeResult;
}[]
//or
const vaildResult: FinishJudgesArgs
```
这个时候只能调用vaildResult了，不能调用args了。另外这是一个数组，应当在this.externalmoduleService.responsefinish内循环处理这个数组。

Handgrip replied:
https://github.com/ThinkSpiritLab/heng-controller/pull/17#discussion_r584066436

协议仓库貌似还没改，那这里就先改成判断`vaildResult.lenght`+仅使用`vaildResult[0]`。

Handgrip replied:
https://github.com/ThinkSpiritLab/heng-controller/pull/17/files#r584228100 别用args数组啦，vaildResult是args筛出来的呀，另外判断一下vaildResult的长度。

Handgrip commented on src/external-module/external-module.service.ts:
```diff
@@ -0,0 +1,132 @@
+                    this.keys.CalbckURLFin,
+                    taskid.toString()
+                )
+            )[0] || "";
```
这边假如是null，应该throw或者continue吧。推荐continue，因为这个时候此任务已经报废了。
上面也是这样。

A0nameless0man replied:
此处如果null了，建议打个Warning

Handgrip commented on src/external-module/external-module.service.ts:
```diff
@@ -0,0 +1,132 @@
+
+    async responsefinish(
+        taskid: number,
+        result: FinishJudgesArgs
```
此处参数类型：
```ts
vaildResult: {
    id: string;
    result: JudgeResult;
}[]
//or
vaildResult : FinishJudgesArgs
```

另外函数内部应当循环处理此数组。
上面也是这样。

revue-starlight replied:
好像已经删除了批量请求，validResult不是数组了，但是协议还没来的及改。。 @A0nameless0man 麻烦确认一下是不是这样

A0nameless0man replied:
批量请求已从设计中移除

Handgrip commented on src/judger/judger.service.ts:
```diff
@@ -36,33 +37,19 @@ export class JudgerService {
-        // if (!infoStr) {
+        // if (
+        //     !(await this.redisService.client.sismember("pendingTask", taskId))
+        // ) {
```
既然上面复制了这里原来的注释，原来的注释其实不动就行。
另外这个pendingTask是debug用的；另外改完之后注释内应该花括号不匹配。

Handgrip replied:
我好像没法resolve这段对话😲

revue-starlight replied:
对唔住..我也不太理解

Handgrip commented on src/external-module/external-module.service.ts:
```diff
@@ -0,0 +1,132 @@
+            taskid,
+            req
+        );
+        await this.redisService.client.hmset(
```
hset
上面也是

Handgrip commented on src/external-module/external-module.service.ts:
```diff
@@ -0,0 +1,132 @@
+            seq: seq,
+            time: moment().format("YYYY-MM-DDTHH:mm:ssZ")
+        };
+        return inreq;
```
这里返回类型CreateJudgeRequest，下面getJudgeINFO()返回CreateJudgeArgs，写入读出的类型都不一样。
所以这里返回inreq.body.method就行。其他多余的都可以删掉，比如this.keys.Seq。

revue-starlight replied:
因为是直接传Args，这里会有较大的删除

Handgrip commented on src/external-module/external-module.service.ts:
```diff
@@ -0,0 +1,132 @@
+
+    // 创建评测任务
+    async createjudge(req: external.CreateJudgeRequest): Promise<void> {
+        await this.redisService.client.incr(this.keys.Seq);
```
可能是重复or多余语句

Handgrip replied:
猜测这里为this.keys.Taskseq，如果是请使用redis事务（this.redisService.client.multi()）把下面一行和这一行连起来。

Handgrip replied:
incr返回加一后该键的值

Handgrip commented on src/external-module/external-module.controller.ts:
```diff
@@ -0,0 +1,15 @@
+
+    // 分发任务
+    @Post("/v1/judges")
+    async CreateJudgeReq(@Body() req: CreateJudgeRequest): Promise<void> {
```
class-vaildator还没加

A0nameless0man replied:
需要的元数据尚未添加至协议仓库

A0nameless0man commented on docs/外部通信模块.md:
```diff
@@ -0,0 +1,21 @@
+
+## 使用的 redis 键值对
+
+'externalmodule:calbckurl:upd' : taskid - url 用于存放回调任务状态的url
```
此处记录的键值与代码中不一致，包括下面几行，请修正

A0nameless0man commented on src/external-module/external-module.service.ts:
```diff
@@ -0,0 +1,132 @@
+    FinishJudgesArgs,
+    UpdateJudgesArgs
+} from "heng-protocol/internal-protocol/ws";
+import * as external from "heng-protocol/external-protocol";
```
请明确所引用的内容，这里好像只引用了一个类型。
`import { CreateJudgeRequest} from "heng-protocol/external-protocol";`

revue-starlight replied:
明确的话会和Internal-protocol里的CreateJudgeRequest冲突，有没有什么好方法解决呢？。。。

revue-starlight replied:
明确的话会和Internal-protocol里的CreateJudgeRequest冲突，有没有什么好方法解决呢？。。。

revue-starlight replied:
明白了，不需要做协议转换以后就没有这个问题了

A0nameless0man replied:
`import { CreateJudgeRequest as XXX} from "heng-protocol/external-protocol";`

A0nameless0man commented on src/external-module/external-module.service.ts:
```diff
@@ -0,0 +1,132 @@
+                args: exreq
+            },
+            seq: seq,
+            time: moment().format("YYYY-MM-DDTHH:mm:ssZ")
```
`new Date().toISOString()` 即可

A0nameless0man replied:
不需要添加不必要的依赖

Handgrip replied:
Date不行，之前测试，出来的时间是+0时区，而不是+8

A0nameless0man replied:
难顶，确实

A0nameless0man commented on src/external-module/external-module.service.ts:
```diff
@@ -0,0 +1,132 @@
+    ) {}
+
+    // 协议内外转换
+    async judgereqconvert(
```
这个函数的功能有问题，外部模块需要将请求中的评测信息存储到Redis中，但是将其包装为内部通信的格式这个工作是内部模块进行的。

A0nameless0man commented on src/external-module/external-module.service.ts:
```diff
@@ -0,0 +1,132 @@
+            taskid,
+            JSON.stringify(inreq)
+        );
+        this.logger.log("发送评测任务 id: ${taskid} ");
```
应该是评测任务已进入队列

Handgrip commented on src/external-module/external-module.service.ts:
```diff
@@ -0,0 +1,132 @@
+                    taskid.toString()
+                )
+            )[0] || "";
+        await this.axiosService.post(url, { taskid, state });
```
由上文得知CalbckURLUpd有可能为undefined，这里应该缺少相应判断。

Nugine replied:
CalbckURLUpd 这命名是缩写还是typo？就算缩写也没有这种扣掉两个字母的写法……

revue-starlight replied:
> CalbckURLUpd 这命名是缩写还是typo？就算缩写也没有这种扣掉两个字母的写法……

查阅了资料，callback是cb。。

Handgrip reviewed (changes requested):
另外缺少redis相应key的垃圾清理，比如任务完成后删除CalbckURLUpd，麻烦点的可能包括异常任务的垃圾清理。

A0nameless0man commented:
> 另外缺少redis相应key的垃圾清理，比如任务完成后删除CalbckURLUpd，麻烦点的可能包括异常任务的垃圾清理。

异常任务指哪些情形呢？如果评测机掉线等会重测，任务无法实际执行会产生SE，仍然会报告结果。

Handgrip commented:
> > 另外缺少redis相应key的垃圾清理，比如任务完成后删除CalbckURLUpd，麻烦点的可能包括异常任务的垃圾清理。
> 
> 异常任务指哪些情形呢？如果评测机掉线等会重测，任务无法实际执行会产生SE，仍然会报告结果。

可以认为任务最终如果汇报了结果，那么外部交互中相应key会被清理。
目前会主动丢弃任务的情况有
1. 无法获取judgeInfo，此情况任务id会进入黑名单，再次取出此任务时会从任务队列中删除此id。相应代码位于 https://github.com/ThinkSpiritLab/heng-controller/blob/master/src/scheduler/judge-queue-service/judge-queue-service.service.ts#L78 以下。可考虑扩充84行完成对cbURL的清理。
2. 待发现

revue-starlight committed:
[external-module]

进行了大的修改，实现了要求
1.删除了内外转换协议函数，将CreateJudgeArgs传递给judger
2.添加了任务提交时的时间戳
3.增添了redis中任务缓存的回收功能
4.使用multi处理一些事务
5.修改了键名与文档
6.添加了module,经简单调试可以正常运行

Nugine commented on src/external-module/external-module.controller.ts:
```diff
@@ -0,0 +1,42 @@
+
+    // 分发任务
+    @Post("/judges")
+    async CreateJudgeReq(@Body() Body: CreateJudgeRequest): Promise<any> {
```
不要使用 any，这里应该可以写出类型

Nugine commented on src/external-module/external-module.service.ts:
```diff
@@ -0,0 +1,113 @@
+        taskid: string,
+        state: UpdateJudgesArgs
+    ): Promise<void> {
+        const url: any = (
```
同上，不要使用 any

revue-starlight replied:
嗯

Nugine commented on src/external-module/external-module.service.ts:
```diff
@@ -0,0 +1,113 @@
+    ) {}
+
+    // 创建评测任务
+    async createjudge(req: CreateJudgeRequest): Promise<void> {
```
为什么用两个单词拼起来的命名？js 的类方法命名通常使用小驼峰。

revue-starlight replied:
嗯，已修改

Nugine commented on src/external-module/external-module.service.ts:
```diff
@@ -0,0 +1,113 @@
+    async responsefinish(
+        taskid: string,
+        result: FinishJudgesArgs
+    ): Promise<any> {
```
any...

Nugine commented on src/external-module/external-module.service.ts:
```diff
@@ -0,0 +1,113 @@
+        }
+    }
+
+    async getJudgeINFO(taskId: string): Promise<CreateJudgeArgs> {
```
INFO 为啥大写...

revue-starlight replied:
dbq...不太熟悉风格

Nugine reviewed (changes requested):
1. 命名风格不统一，最好改改。
2. 除非万不得已，不要使用 any。
3. 由于 yarn v1 的 bug，协议仓库不能自动编译，切换为 npm 吧。具体步骤：删掉 yarn.lock，npm install，提交 package-lock.json。
4.  @Flaryer @revue-starlight 移除批量更新后的改动不大，确认解决后就 resolve conversation

如果周末解决了所有问题，我将于下周一晚上合并此 PR。

Handgrip commented on src/external-module/external-module.service.ts:
```diff
@@ -0,0 +1,113 @@
+            req.id,
+            moment().format("YYYY-MM-DDTHH:mm:ssZ")
+        );
+        await mu.exec();
```
官方教程里没说用mu存储multi并且包含多个语句时该怎么写，但是我感觉这里什么语句都不会执行（我也还没测试）。应该是mu=mu.hmset(...);
或者直接链式。
另外为啥还是hmset，下面还有hmget，虽然没什么大影响。

revue-starlight replied:
已修改

Handgrip commented on src/external-module/external-module.service.ts:
```diff
@@ -0,0 +1,113 @@
+            mu.del(this.keys.CBURLUpd, taskid);
+            mu.del(this.keys.CBURLFin, taskid);
+            mu.del(this.keys.TaskTime, taskid);
+            await mu.exec();
```
同上

Handgrip replied:
然后这里是hdel。。。你这样把整个hash键删了啊

Handgrip replied:
mu=mu.hdel()

revue-starlight replied:
惊天大bug!!!!!!!

Handgrip commented on docs/外部通信模块.md:
```diff
@@ -0,0 +1,25 @@
+
+'Ext:UrlFin' : taskid - url 用于存放回调任务结果的url
+
+'Ext:JudgeINFO' : taskid - JSON.stringify(Internal.CreateJudgeRequest) 用于存放评测任务信息
```
是CreateJudgeArgs

Handgrip commented on src/external-module/external-module.service.ts:
```diff
@@ -0,0 +1,113 @@
+        mu.hmset(
+            this.keys.TaskTime,
+            req.id,
+            moment().format("YYYY-MM-DDTHH:mm:ssZ")
```
我认为这种不给人看的时间设置成timestamp就好，Date.now()，方便读取，比较。

Handgrip reviewed (changes requested):
另外貌似TaskTime还没利用到，应该是垃圾清理相关的。

revue-starlight commented:
确实（------------------&nbsp;原始邮件&nbsp;------------------
发件人:&nbsp;&quot;Flaryer&quot;<notifications@github.com&gt;
发送时间:&nbsp;2021年3月7日(星期天) 上午9:43
收件人:&nbsp;&quot;ThinkSpiritLab/heng-controller&quot;<heng-controller@noreply.github.com&gt;;
抄送:&nbsp;&quot;revue-starlight&quot;<204214884@qq.com&gt;;&quot;Mention&quot;<mention@noreply.github.com&gt;;
主题:&nbsp;Re: [ThinkSpiritLab/heng-controller] 外部交互模块 (#17)

revue-starlight commented:
judger那一段代码我不太熟悉，删除了多组评测后，我改的应该不太好。麻烦大家再查看一下。    其他的都作出了修改，如果有不完善的地方还请指正。谢谢各位大佬～。🌸------------------&nbsp;原始邮件&nbsp;------------------
发件人:&nbsp;&quot;Nugine&quot;<notifications@github.com&gt;
发送时间:&nbsp;2021年3月5日(星期五) 晚上9:21
收件人:&nbsp;&quot;ThinkSpiritLab/heng-controller&quot;<heng-controller@noreply.github.com&gt;;
抄送:&nbsp;&quot;revue-starlight&quot;<204214884@qq.com&gt;;&quot;Mention&quot;<mention@noreply.github.com&gt;;
主题:&nbsp;Re: [ThinkSpiritLab/heng-controller] 外部交互模块 (#17)

Commit:
1.修改了键名
2.将any修改为固定的类型名称
3.yarn修改为npm
4.修改了 bug: responsefinish中直接删除了

Commit:
1.完成了对键名，一些命名规则的修改
2.yarn切换为npm
3.修改了bug:redis删除时del -> hdel
4.doc修改了相应内容
5.judge应对协议取消批量评测做出了部分修改

Commit:
1.完成了对键名，一些命名规则的修改 2.yarn切换为npm 3.修改了bug:redis删除时del -> hdel 4.doc修改了相应内容 5.judge应对协议取消批量评测做出了部分修
改

Commit:
对命名方式做出了修改

A0nameless0man commented on docs/外部通信模块.md:
```diff
@@ -4,22 +4,22 @@
 
 1. 从用户端接收评测任务
-   将 external.CreateJudgeRequest 从外部通信协议格式转换为内部通信协议格式。并将其转化为 JSON 字符串，放入 redis 队列中。塞入评测队列。等待评测结果回调
+   将 external.CreateJudgeRequest 从外部通信协议格式存入 redis 中。并将其转化为 JSON 字符串，放入 redis 队列中。塞入评测队列。等待评测结果回调
```
将其转化为 JSON 字符串，放入hset，将id交给任务队列模块吧。

A0nameless0man commented on src/external-module/external-module.controller.ts:
```diff
@@ -12,31 +14,36 @@ export class ExternalModuleController {
-    //     this.externalmoduleService.responsefinish(Args[0].id,Args);
-    // }
+    //用于debug,此处debug的作用为模拟更新结果的函数
+    @Get("/test/:id")
```
建议换个人类可读的路径，比如 `/test/reportFinish/:id` 或 `/test/reportFinish/?id=:id`

A0nameless0man commented on src/external-module/external-module.controller.ts:
```diff
@@ -12,31 +14,36 @@ export class ExternalModuleController {
+    @Post("/testurl")
+    async testurl(@Body() Body: CreateJudgeRequest): Promise<void> {
+        console.log("收到url一份,url的body信息为：");
+        console.log(Body);
```
请用nestjs提供的logger，请使用ts提供的字符串拼接方法

A0nameless0man commented on src/external-module/external-module.service.ts:
```diff
@@ -38,63 +40,74 @@ export class ExternalModuleService {
+        mu = mu.hmset(this.keys.JudgeInfo, req.id, JSON.stringify(Args)); // 这里好像跑起来没问题...
+        mu = mu.hmset(this.keys.CBURLUpd, req.id, req.callbackUrls.update);
+        mu = mu.hmset(this.keys.CBURLFin, req.id, req.callbackUrls.finish);
+        mu = mu.hmset(this.keys.TaskTime, req.id, Date.now());
```
确实没问题，但还是建议链式

Nugine replied:
hmset 可以一次性设置多个键值对

https://redis.io/commands/hmset

A0nameless0man commented on src/external-module/external-module.service.ts:
```diff
@@ -38,63 +40,74 @@ export class ExternalModuleService {
         await mu.exec();
         await this.judgequeueService.push(String(req.id));
         this.logger.log(`评测任务已进入队列 id: ${req.id} `);
+        req.id;
```
这一行是干啥？

A0nameless0man commented on src/external-module/external-module.service.ts:
```diff
@@ -38,63 +40,74 @@ export class ExternalModuleService {
+            mu.hdel(this.keys.JudgeInfo, taskid);
+            mu.hdel(this.keys.CBURLUpd, taskid);
+            mu.hdel(this.keys.CBURLFin, taskid);
+            mu.hdel(this.keys.TaskTime, taskid);
```
同样，建议链式调用

A0nameless0man commented on src/judger/judger.service.ts:
```diff
@@ -124,60 +124,49 @@ export class JudgerService {
-        args.forEach(({ id }) => {
-            mu = mu.sismember(wsId + WsOwnTaskSuf, id);
-        });
+        mu = mu.sismember(wsId + WsOwnTaskSuf, args.id);
```
单个命令不需事务

Nugine commented on src/external-module/external-module.controller.ts:
```diff
@@ -0,0 +1,49 @@
+
+    // 分发任务
+    @Post("/judges")
+    async CreateJudgeReq(
```
一般用小驼峰命名，首字母小写。

revue-starlight commented:
嗯...hmset好像只能批量键值，在不同的key里不能用...我这里要不还是改为hset吧------------------&nbsp;原始邮件&nbsp;------------------
发件人:&nbsp;&quot;Nugine&quot;<notifications@github.com&gt;
发送时间:&nbsp;2021年3月8日(星期一) 中午11:45
收件人:&nbsp;&quot;ThinkSpiritLab/heng-controller&quot;<heng-controller@noreply.github.com&gt;;
抄送:&nbsp;&quot;revue-starlight&quot;<204214884@qq.com&gt;;&quot;Mention&quot;<mention@noreply.github.com&gt;;
主题:&nbsp;Re: [ThinkSpiritLab/heng-controller] 外部交互模块 (#17)

revue-starlight committed:
[external]

1.createJudgeReq小驼峰命名，首字母小写。
2.删除了多余行req.id
3.采用链式处理事务，因为不太会hmset的多次设置用法
4.用nestjs提供的logger
5.修改doc中创建评测任务的接口说明：将其转化为 JSON字符串，放入hset，将id交给任务队列模块吧。

revue-starlight committed:
对judger内容做出修整

revue-starlight committed:
[external module]

1. 将链式的exec接到链式上，语法规范化
2. 规范url
3. 规范了doc

revue-starlight committed:
修复了一些细节问题，使其可以和client交互运行

Handgrip commented on src/external-module/external-module.service.ts:
```diff
@@ -0,0 +1,126 @@
+@Injectable()
+export class ExternalModuleService {
+    private readonly logger = new Logger("ExternalModuleService");
+    public readonly keys = {
```
建议static，不然别的模块读不到

A0nameless0man replied:
其它模块也不该读这玩意啊

A0nameless0man replied:
其实，其它模块不应该读这些。

Handgrip replied:
方便以后修改吧，就像现在内部通信读了评测机池的键

revue-starlight replied:
等到要读的时候再改也不迟

Handgrip commented on src/external-module/external-module.service.ts:
```diff
@@ -0,0 +1,126 @@
+            )
+        )[0];
+        if (url == null) {
+            this.logger.warn(`未找到更新状态id ${taskid} 的回调url`);
```
return;

Handgrip replied:
这个评论忽略，忘记删了，不用改。

Handgrip commented on src/external-module/external-module.service.ts:
```diff
@@ -0,0 +1,126 @@
+                this.keys.CBURLUpd,
+                taskid.toString()
+            )
+        )[0];
```
用hget好像返回的直接是value，不用加 [0]

Handgrip commented on src/external-module/external-module.service.ts:
```diff
@@ -0,0 +1,126 @@
+                this.keys.CBURLFin,
+                taskid.toString()
+            )
+        )[0];
```
同上

Handgrip commented on src/external-module/external-module.service.ts:
```diff
@@ -0,0 +1,126 @@
+                .hdel(this.keys.CBURLFin, taskid)
+                .hdel(this.keys.TaskTime, taskid)
+                .exec();
+            this.logger.log(`已返回评测任务id: ${taskid} 的结果`);
```
删除console.log()

Handgrip commented on src/judger/judger.service.ts:
```diff
@@ -125,18 +116,17 @@ export class JudgerService {
+            return;
+        }
+
         // TODO 通知外部系统
```
换为自己的函数。

Handgrip commented on src/judger/judger.service.ts:
```diff
@@ -145,29 +135,20 @@ export class JudgerService {
+            return;
+        }
 
         // TODO 通知外部系统
```
换为自己的函数。

revue-starlight commented:
奇怪，我这个在本地改过了啊。...------------------&nbsp;原始邮件&nbsp;------------------
发件人:&nbsp;&quot;Flaryer&quot;<notifications@github.com&gt;
发送时间:&nbsp;2021年3月9日(星期二) 下午2:39
收件人:&nbsp;&quot;ThinkSpiritLab/heng-controller&quot;<heng-controller@noreply.github.com&gt;;
抄送:&nbsp;&quot;revue-starlight&quot;<204214884@qq.com&gt;;&quot;Mention&quot;<mention@noreply.github.com&gt;;
主题:&nbsp;Re: [ThinkSpiritLab/heng-controller] 【功能模块开发】外部交互模块 (#17)

revue-starlight commented:
push不上去力，，，------------------&nbsp;原始邮件&nbsp;------------------
发件人:&nbsp;&quot;Flaryer&quot;<notifications@github.com&gt;
发送时间:&nbsp;2021年3月9日(星期二) 下午2:39
收件人:&nbsp;&quot;ThinkSpiritLab/heng-controller&quot;<heng-controller@noreply.github.com&gt;;
抄送:&nbsp;&quot;revue-starlight&quot;<204214884@qq.com&gt;;&quot;Mention&quot;<mention@noreply.github.com&gt;;
主题:&nbsp;Re: [ThinkSpiritLab/heng-controller] 【功能模块开发】外部交互模块 (#17)

revue-starlight committed:
lint&format

Handgrip commented on src/external-module/external-module.service.ts:
```diff
@@ -0,0 +1,131 @@
+            .hset(this.keys.CBURLFin, req.id, req.callbackUrls.finish)
+            .hset(this.keys.TaskTime, req.id, Date.now())
+            .exec();
+        await this.judgequeueService.push(String(req.id));
```
这里不用string()了，上面有了。

Handgrip commented on src/external-module/external-module.service.ts:
```diff
@@ -0,0 +1,131 @@
+                    `返回taskid:${taskid}评测结果失败，该任务回调url: ${url} 无法正常连接`
+                );
+            }
+            this.cleanJudge(taskid);
```
建议这一行放在if else外面，防止跳过清理垃圾。

revue-starlight replied:
好

Handgrip commented on src/judger/judger.gateway.ts:
```diff
@@ -659,10 +659,10 @@ export class JudgerGateway implements OnGatewayInit, OnGatewayConnection {
+                    //if (body.output !== undefined) {
+                    resolve(body.output);
+                    //}
+                    //reject(new Error("Empty Response"));
```
改成if else

Handgrip commented on src/external-module/external-module.service.ts:
```diff
@@ -0,0 +1,131 @@
+    // 评测任务回调
+    async responseUpdate(
+        taskid: string,
+        state: UpdateJudgesArgs
```
仓库已经改了，这里state类型为JudgeState，调用时是args.state，下面finish函数也要改

revue-starlight committed:
修整

revue-starlight committed:
修改补充

Handgrip commented on src/external-module/external-module.service.ts:
```diff
@@ -0,0 +1,118 @@
+            } catch (error) {
+                this.logger.warn(
+                    `返回taskid:${taskid}评测结果失败，该任务回调url: ${url} 无法正常连接`
+                );
```
有个问题：连接失败时应当throw，这将导致RPC失败，按照计划 @A0nameless0man 评测机那里会再次回报结果。因此这里不仅打log，还要throw。上面回报状态我不知道要不要这样，可能导致评测机那里结果回报了但是一堆状态未回报（非洲评测机）？问问 @A0nameless0man 怎么搞。

Nugine replied:
@A0nameless0man 评测机对这种情况是怎么处理的

A0nameless0man replied:
这里回报失败我觉得不应当向评测机通报，应当由外部模块尽最大努力（指若干次重试）后放弃。

A0nameless0man replied:
同时，对评测机来说，这个上报的rpc调用应当尽快返回，因为评测机不关心对外部系统的调用是否成功。

revue-starlight committed:
修改了bug

judgerservice进行了修改
增加了指数退避

Nugine commented:
最后再确认一下，review 都通过后就合并。
